apps/common-app/src/examples/DrumMachine/usePlayer.tsx

```diff
@@ -0,0 +1,157 @@
+import { AudioContext } from 'react-native-audio-api';
+import { useSharedValue } from 'react-native-reanimated';
+import { useRef, useState, useCallback, useLayoutEffect } from 'react';
+
+import type { InstrumentName, Pattern } from './types';
+import { Kick, HiHat, Clap } from '../SharedUtils';
+import { numBeats } from './constants';
+
+interface PlayerOptions {
+  bpm: number;
+  patterns: Pattern[];
+  notesPerBeat: number;
+}
+
+function r<T>(ref: React.MutableRefObject<T>) {
+  return ref.current;
+}
+
+export default function usePlayer(options: PlayerOptions) {
+  const { bpm, patterns, notesPerBeat } = options;
+
+  const [isPlaying, setIsPlaying] = useState(false);
+
+  const bpmRef = useRef(bpm);
+  const patternsRef = useRef(patterns);
+  const isPlayingRef = useRef(isPlaying);
+  const notesPerBeatRef = useRef(notesPerBeat);
+
+  const progressSV = useSharedValue(0);
+  const currentBeat = useSharedValue(0);
+  const playingNotes = useSharedValue<boolean[]>(
+    Array(patterns.length).fill(false)
```

Comment:
ideally it should use map `{ [instrumentName]: boolean }` the order or instruments will be irrelevant